aten/src/ATen/native/cuda/DilatedMaxPool2d.cu

```diff
@@ -129,6 +210,100 @@ __global__ void MaxPoolBackward(const int nthreads, const scalar_t* top_diff,
+
+  int iH = (height + gridDim.z-1) / gridDim.z;
+  int iW = (width + gridDim.y-1) / gridDim.y;
+  int istartH = threadIdx.z + blockIdx.z*iH;
+  int iendH = ::min(istartH+iH, height);
+  int istartW = threadIdx.y + blockIdx.y*iW;
+  int iendW = ::min(istartW+iW, width);
+
+  for (int ih = istartH; ih < iendH; ih+=blockDim.z) {
+    for (int iw = istartW; iw < iendW; iw+=blockDim.y) {
+      int phstart, phend, pwstart, pwend;
+      if (stride_h == 1) {
+        phstart =
+            (ih + pad_h < ((kernel_h - 1) * dilation_h + 1)) ? 0 : (ih + pad_h - ((kernel_h - 1) * dilation_h + 1))  + 1;
+        phend = min((ih + pad_h)  + 1, pooled_height);
+      } else if (stride_h == 2) {
+        phstart =
+            (ih + pad_h < ((kernel_h - 1) * dilation_h + 1)) ? 0 : (ih + pad_h - ((kernel_h - 1) * dilation_h + 1)) / 2  + 1;
+        phend = min((ih + pad_h) / 2  + 1, pooled_height);
+      } else {
+        phstart =
+            (ih + pad_h < ((kernel_h - 1) * dilation_h + 1)) ? 0 : (ih + pad_h - ((kernel_h - 1) * dilation_h + 1)) / stride_h  + 1;
+        phend = min((ih + pad_h) / stride_h  + 1, pooled_height);
+      }
+      if (stride_w == 1) {
+        pwstart =
+            (iw + pad_w < ((kernel_w - 1) * dilation_w + 1)) ? 0 : (iw + pad_w - ((kernel_w - 1) * dilation_w + 1)) + 1;
+        pwend = min((iw + pad_w) + 1, pooled_width);
+      } else if (stride_w == 2) {
+        pwstart =
+            (iw + pad_w < ((kernel_w - 1) * dilation_w + 1)) ? 0 : (iw + pad_w - ((kernel_w - 1) * dilation_w + 1)) / 2 + 1;
+        pwend = min((iw + pad_w) / 2 + 1, pooled_width);
+      } else {
+        pwstart =
+            (iw + pad_w < ((kernel_w - 1) * dilation_w + 1)) ? 0 : (iw + pad_w - ((kernel_w - 1) * dilation_w + 1)) / stride_w + 1;
+        pwend = min((iw + pad_w) / stride_w + 1, pooled_width);
+      }
+
+      if ((phstart + 1 != phend) || (pwstart + 1 != pwend)) {
+        for(int oh = phstart; oh < phend; ++oh) {
+          for(int ow = pwstart; ow < pwend; ++ow) {
+            const int64_t* ptr_top_mask = top_mask + oh * out_stride_h + ow * out_stride_w;
+            for (int c = threadIdx.x; c < channels; c += blockDim.x) {
+              if (ptr_top_mask[c] == ih * width + iw) {
+                out_cached[c] += ScalarConvert<accscalar_t, scalar_t>::to(top_diff[oh * out_stride_h + ow * out_stride_w + c]);
+              }
+            }
+          }
+        }
+      } else {
+        int idx = phstart * out_stride_h + pwstart * out_stride_w;
+        const int64_t* ptr_top_mask = top_mask + idx;
+        for (int c = threadIdx.x; c < channels; c += blockDim.x) {
+          if (ptr_top_mask[c] == ih * width + iw) {
+            out_cached[c] += ScalarConvert<accscalar_t, scalar_t>::to(top_diff[phstart * out_stride_h + pwstart * out_stride_w + c]);
+          }
+        }
+      }
+
+      scalar_t *ptr_bottom_diff = bottom_diff + (ih * width + iw) * channels;
```

Comment:
Arguably out_cached is not really beneficial when we special case `(phstart + 1 == phend) && (pwstart + 1 == pwend)`

Reply:
Apology that last comment is confusing.
in this special case here, we are not iterating on output data, but a simple 1 to 1 mapping. Really there's no benefit of going from `global -> register -> shared memory -> register -> global` here when we can do `global -> register -> global`

We should just write `bottom_diff[yyy] = top_diff[xxx]` directly.

Reply:
Fixed